common/context/src/main/java/io/helidon/common/context/Context.java

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * A registry for context objects. Enables instance localization between several <i>services / components / ...</i> integrated in
+ * a particular known scope. ContextualRegistry instance is intended to be associated with a scope aware object such as
+ * WebServer, ServerRequest or ClientRequest.
+ *
+ * <p>Context contains also a notion of <i>classifiers</i>. Classifier is any object defining additional <i>key</i> for registered
+ * objects. To obtain such registered object, the same classifier (precisely, any equal object) has to be used.
+ *
+ * <p>Classifiers can be used as follows:<ol>
+ * <li>As an additional identifier for registered objects of common types, like a {@link String}, ...<br>
+ * <pre>{@code
+ * // User detail provider service
+ * registry.register("NAME_PARAM_ID", "Smith");
+ * registry.register("GENDER_PARAM_ID", "male");
+ * ...
+ * // User consumer service
+ * String name = registry.get("name", String.class);
+ * }</pre></li>
+ * <li>As an access control mechanism where only owners of the classifier can retrieve such contextual instance.<br>
+ * <pre>{@code
+ * // In some central security service.
+ * registry.register(securityFrameworkInternalInstance, new AuthenticatedInternalIdentity(...));
+ * ...
+ * // In some authorization filter known by a central security service
+ * AuthenticatedInternalIdentity auth = registry.get(securityFrameworkInternalInstance, AuthenticatedInternalIdentity.class);
+ * }</pre></li>
+ * </ol>
+ */
+public interface Context {
+
+    /**
+     * Creates a new empty instance.
+     *
+     * @return new instance
+     */
+    static Context create() {
+        return new ListContext();
+    }
+
+    /**
+     * Creates a new empty instance backed by its parent read-through {@link Context}.
+     *
+     * <p>Parent {@code registry} is used only for get methods and only if this registry doesn't have registered required type.
+     *
+     * @param parent a parent registry
+     * @return new instance
+     */
+    static Context create(Context parent) {
+        return new ListContext(parent);
+    }
+
+    /**
+     * Register a new instance.
+     *
+     * @param instance an instance to register
+     * @param <T>      a type of the registered instance
+     * @throws NullPointerException if the registered object is {@code null}
+     */
+    <T> void register(T instance);
+
+    /**
+     * Register a new instance using a provided supplier. The supplier is guarantied to be called at most once when it's
+     * requested by the {@link #get(Class)} method. The returned value is then registered and the supplier is never used again.
+     *
+     * @param type     a type of supplied instance
+     * @param supplier a supplier of the instance to register
+     * @param <T>      a type of supplied object
+     * @throws NullPointerException if the {@code type} or the {@code supplier} is {@code null}
+     */
+    <T> void supply(Class<T> type, Supplier<T> supplier);
```

Comment:
If you can do so without conflicts, I think you should use "register" instead of "supply" (everywhere). As it is, you're naming the how, not the what, and it is a little confusing, especially when the first word of the javadoc is "Register".

Reply:
Never mind; I see below that you hoisted this from a public API.